### PR TITLE
[Driver] Fix detection of libc++ with empty sysroot.

### PR DESCRIPTION
**Pull request by sam-mccall**
b1e3cd1d79443603dc003441e07cdd8d30bb7f26 dropped the leading slash here,
presumably unintentionally.

(I don't understand Driver well enough to know how/where this is supposed
to be tested, but it broke clangd on any project that uses the
system-installed -stdlib=libc++ on a standard debian install)

### CODE REVIEW

**Comment by smeenai** [comment]
Sorry about the breakage.

The `sys::path::append` behavior here is surprising. I think it'd be better to change the computation of `SysRoot` in the function above (line 3102, which I can't comment because of GitHub, sigh) to default to `llvm::sys::path::get_separator()` if it's empty, to handle Windows slashes as well as be less confusing to read.

This is tricky to test because it requires no sysroot to be passed, but we might be able to do so with a VFS overlay. I'm okay adding the test afterwards though so we get the breakage fixed first, and I can also look into the test myself if you'd like.

**Comment by sam-mccall** [comment]
Thanks for the quick response!

> The `sys::path::append` behavior here is surprising. I think it'd be better to change the computation of `SysRoot` in the function above (line 3102, which I can't comment because of GitHub, sigh) to default to `llvm::sys::path::get_separator()` if it's empty, to handle Windows slashes as well as be less confusing to read.

That sounds sensible to me in principle, but not trivial: presumably the overrides of `Toolchain::getSysroot()` should have a consistent contract. There are 6 overrides, where some explicitly return the empty string, and 25 callers some of which do things like `computeSysRoot() + "/include/c++/"` (RISCVToolchain).
I feel it's likely I'll break something on a platform I don't understand before getting back to a good state :-)

> This is tricky to test because it requires no sysroot to be passed, but we might be able to do so with a VFS overlay. I'm okay adding the test afterwards though so we get the breakage fixed first, and I can also look into the test myself if you'd like.

I can try to add something to the unittest. Looks like in the current form I also need to change the lit test back to hard-coding `/` for this one separator.
(I'm out of time tonight but will pick it up tomorrow)

**Comment by smeenai** [comment]
> Thanks for the quick response!
> 
> > The `sys::path::append` behavior here is surprising. I think it'd be better to change the computation of `SysRoot` in the function above (line 3102, which I can't comment because of GitHub, sigh) to default to `llvm::sys::path::get_separator()` if it's empty, to handle Windows slashes as well as be less confusing to read.
> 
> That sounds sensible to me in principle, but not trivial: presumably the overrides of `Toolchain::getSysroot()` should have a consistent contract. There are 6 overrides, where some explicitly return the empty string, and 25 callers some of which do things like `computeSysRoot() + "/include/c++/"` (RISCVToolchain). I feel it's likely I'll break something on a platform I don't understand before getting back to a good state :-)

I just meant to change https://github.com/llvm/llvm-project/blob/5eaa5312e7943e23155da4f0fbf07b55a200fc60/clang/lib/Driver/ToolChains/Gnu.cpp#L3102 to something like (untested):

```
std::string SysRoot = computeSysRoot();
if (SysRoot.empty())
  SysRoot = llvm::sys::path::get_separator();
```

It's a local fix, but all the code here is pretty inconsistent anyway, so I don't feel too bad about it.

**Comment by sam-mccall** [comment]
> I just meant to change [...] It's a local fix, but all the code here is pretty inconsistent anyway, so I don't feel too bad about it.

Oops, of course. Done & I worked out the testing.

File: clang/unittests/Driver/ToolChainTest.cpp [hunk 1]
```diff
@@ -316,6 +318,52 @@ TEST(ToolChainTest, VFSSolarisMultiGCCInstallation) {
...
+  llvm::ListSeparator Sep(" ");
+  for (const char *Arg : C.getArguments()) {
+    Args += Sep;
+    Args += Arg;
+  }
+  if (is_style_windows(llvm::sys::path::Style::native))
+    std::replace(Args.begin(), Args.end(), '\\', '/');
+  if (llvm::StringRef(Args).contains(Substr))
+    return true;
+  *result_listener << "whose args are '" << Args << "'";
```
Comment (smeenai) [inline clang/unittests/Driver/ToolChainTest.cpp:333]: Not familiar with the unit testing set up ... do you need the "whose args are" bit?

**Comment by smeenai** [comment]
LGTM, thanks!

**Review by smeenai — APPROVED** [review]

File: clang/unittests/Driver/ToolChainTest.cpp (hunk 1 above)
Comment (sam-mccall) [reply clang/unittests/Driver/ToolChainTest.cpp:333]: Yeah, this is printed only when the matcher fails. It's needed for a decent error message as the job itself has no useful printer.

Comment (smeenai) [reply clang/unittests/Driver/ToolChainTest.cpp:333]: Cool, sounds good.

**Comment by vvereschaka** [comment]
@sam-mccall ,
your 7ca8c21af36acb117529e797b3d36e85a286ca47 commit breaks `Clang-Unit :: Driver/./ClangDriverTests.exe/ToolChainTest/VFSGnuLibcxxPathNoSysroot` (`Clang-Unit::67`) test on the windows toolchain builders
* https://lab.llvm.org/buildbot/#/builders/119/builds/15230
* https://lab.llvm.org/buildbot/#/builders/60/builds/14031 

The problem remains for more than 3 days already. Would you fix it as soon as it possible or revert your changes?

**Comment by sam-mccall** [comment]
Sorry, there were no notifications, I guess the bot was already red.

I'm not really sure what to do other than disable the test on these bots - I don't know why it would fail on win/arm (it seems to pass on win/x86), and I don't have access to such a machine.
This behavior isn't particularly relevant on windows but is critical on linux. This isn't a regression, there was just previously no testing of it.

**Comment by sam-mccall** [comment]
... and as soon as I said that, I found something: looks like these are maybe building with `DEFAULT_SYSROOT=C:/buildbot/.arm-ubuntu`, which breaks the no-sysroot behavior we're testing here.

I'll see if I can override with `--sysroot=`

**Comment by asl** [comment]
@sam-mccall As far as I can see, the bot was green before this change: https://lab.llvm.org/buildbot/#/builders/119/builds/15229

Have you not received email from the buildbot?

**Comment by vvereschaka** [comment]
>edit: hopefully good after https://github.com/llvm/llvm-project/commit/5aa3338930d15a59dd6ddbd36fc09ffa6610ca72

the test has been passed locally on the builder with this commit. Thank you for the fix.

**Pull request merged**